src/scene/shader-lib/programs/basic.js

```diff
@@ -9,10 +9,10 @@ import {
 } from '../../constants.js';
 import { ShaderPass } from '../../shader-pass.js';
 
-import { begin, end, fogCode, skinCode } from './common.js';
+import { ShaderGenerator } from './shader-generator.js';
 
-const basic = {
-    generateKey: function (options) {
+class ShaderGeneratorBasic extends ShaderGenerator {
```

Comment:
What's the purpose of this extends?

Reply:
To give all those classes base class / shared type information.

Reply:
I don't see that used anywhere?

The structure of this whole class hierarchy is very non-standard (and for me confusing).

ShaderGenerator isn't a real base class because it defines no base functionality. Rather it is a set of static helper functions.

Then ShaderGeneratorBasic and the rest implement what looks like a consistent set of functions (which presumably program-libraray expects, but never defines).

Ultimately the structure/interface of ShaderGenerator isn't defined anywhere and is left as implicit.

I don't think there needs to be any inheritance here and probably program-library should define what structure a 'generator' takes. Each implementation of generator just implements that interface.

Reply:
At some point I'm going to add type information to the program-library, and this is the type of the generator. Ideally this would be an interface, but unfortunately JS does not have a concept of this. But this is the only way to assign a single (base) type to multiple classes.

Reply:
If program-library just uses duck typing on generators, why do generators need a class hierarchy?

Reply:
I added some JSDocs to the program-libray that use the base class type - type safety is where we're going to with the engine changes.